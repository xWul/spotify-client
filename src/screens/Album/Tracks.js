/* LIBS */
import React, { Component } from 'react'
import styled from 'styled-components'
/* API */
import SpotifyWrapper from '../../api/index'
import { TOKEN } from '../../api/config'

const Content = styled.div`
  background-color: rgba(0, 0, 0, 0.5 );
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;
  border: 2px solid #FFF;
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 1;
  transition: all 0.3s ease;
  z-index: 110;
`

const Close = styled.span`
  cursor: pointer;
  display: block;
  padding: 10px;
  float: right;
  color: #FFF;
  width: 100%;
  text-align: right;
  z-index: 120;
`
const List = styled.ul`
  width: 100%;
  height: 89%;
  position: absolute;
  overflow: auto;
  margin-top: 8%;
  background-color: #FFF;
`
const Card = styled.li`
  display: flex;
  align-items: center;
  padding: 5px;
  border-radius: 3px;
  img {
    width: 45px;
    height: 45px;
  }
  p {
    padding: 0px 5px;
  }
`
const Load = styled.p`
  color: #FFF;
  text-align: center;
  width: 100%;
`

const Track = ({ track }) => {
  const { name } = track
  return (
    <Card>
      <p>{name}</p>
    </Card>
  )
}

const TracksList = ({ tracks }) => {
  return (
    tracks.items
      ? <List>
        {
          tracks.items.map(track => {
            return <Track track={track} key={track.id} />
          })
        }
      </List>
      : <Load>No tracks found.</Load>
  )
}

const Loading = () => {
  return <Load>Loading...</Load>
}

class Tracks extends Component {
  state = { tracks: [], isLoading: false }

  componentDidMount() {
    const { id } = this.props
    this.setState({ isLoading: true })
    const spotify = new SpotifyWrapper({
      token: TOKEN
    })
    spotify.album.getAlbum(id)
      .then(response => {
        this.setState({ tracks: response.tracks, isLoading: false })
      })
  }

  render() {
    const { onClose } = this.props
    const { tracks, isLoading } = this.state
    return (
      <Content>
        <Close onClick={onClose}>Close</Close>
        {tracks && <TracksList
          tracks={tracks} />}
        {isLoading && <Loading />}
      </Content>
    )
  }
}

export default Tracks
