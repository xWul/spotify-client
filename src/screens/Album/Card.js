/* LIBS */
import React, { Component } from 'react'
import styled from 'styled-components'
import { screens } from '../../theme/globalStyle'
import { setFavorite, removeFavorite, getFavorites } from '../../utils'
/* COMPONENTS */
import LikeButton from '../../components/LikeButton'
import Tracks from './Tracks'

const Card = styled.div`
  font-family: 'Montserrat', sans-serif;
  position: relative;
  background-color: #FFF;
  color: #333;
  flex-wrap: wrap;
  display: flex;
  margin: 1.5%;
  width: 30%;
  @media (max-width: ${screens.medium}) {
    width: 47%;
  }
  @media (max-width: ${screens.small}) {
    width: 100%;
    margin: 2% 0px;
  }
`
const Picture = styled.div`
  width: 100%;
  height: 200px;
  background: ${props => props.images[0] ? `url(${props.images[0].url})` : '#CCC'};
  background-position: center;
  background-size: cover;
  cursor: pointer;
`
const Title = styled.h3`
  font-size: 20px;
  padding: 7px 0; 
`

const Header = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
`

const CardInfo = styled.div`
  padding: 10px; 
  font-size: 14px;
  width: 100%;
  min-height: 140px;
`
const AvalLabel = styled.span`
  color: ${props => props.aval ? 'green' : '#CCC'};
`

const Artists = ({items}) => { 
  return items.length >= 3 
    ? <p>Various artists</p> 
    : <p>{items.map(artist => artist.name).join(', ')}</p>
}

const Availability = ({availabilities}) => {
  return availabilities.includes('BR') 
    ? <AvalLabel aval={true}>Available</AvalLabel> 
    : <AvalLabel aval={false}>Not Available</AvalLabel>
}

class AlbumCard extends Component {
  state = { 
    isFavorite: false,
    showTracks: false 
  }

  componentDidMount() {
    const { id } = this.props.album
    const favorites = getFavorites()['albums'] || []
    if (favorites.includes(id)) {
      this.setState({isFavorite: true})
    }
  }

  toggleFavorite = (id) => {
    this.setState(({ isFavorite }) => ({
      isFavorite: !isFavorite
    }), () => {
      const {isFavorite} = this.state
      isFavorite 
        ? setFavorite('albums', id) 
        : removeFavorite('albums', id)
    })
  }

  toggleTracks = () => {
    this.setState(({ showTracks }) => ({
      showTracks: !showTracks
    }))
  }

  render() {
    const { id, name, images, artists, available_markets } = this.props.album
    const { isFavorite, showTracks } = this.state
    return (
      <Card>
        {showTracks && <Tracks id={id} onClose={this.toggleTracks} />}
        <Picture images={images} onClick={this.toggleTracks} />
        <CardInfo>
          <Header>
            <Title>{name}</Title>
            <LikeButton
              onClick={() => this.toggleFavorite(id)}
              isChecked={isFavorite}/>
          </Header>
          <Artists items={artists} />
          <Availability availabilities={available_markets}/>
        </CardInfo>
      </Card>
    )
  }
}

export default AlbumCard
