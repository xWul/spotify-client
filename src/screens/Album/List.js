/* LIBS */
import React from 'react'
import styled from 'styled-components'
/* COMPONENTS */
import Card from './Card'

const Container = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
`

const AlbumsList = ({ items }) => {
  return (
    <Container>
      {
        items.map(item => {
          return <Card key={item.id} album={item} />
        })
      }
    </Container>
  )
}

export default AlbumsList
