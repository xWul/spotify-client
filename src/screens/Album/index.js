/* LIBS*/
import React, { Component } from 'react'
import styled from 'styled-components'
import { withRouter } from 'react-router-dom'
import { screens } from '../../theme/globalStyle'
/* API */
import SpotifyWrapper from '../../api/index'
import { TOKEN } from '../../api/config'
/* COMPONENTS */
import Alert from '../../components/Alert'
import Search from '../../components/Search'
import Loader from '../../components/Loader'
import List from './List'

const Container = styled.main`
  min-height: 100vh;
  background-color: #1D1D1D;
  padding: 40px 10%;
  @media(max-width: ${screens.small}) {
    padding: 40px 5% 100px 5%;
  }
`

class Album extends Component {
  state = {
    albums: [],
    isLoading: false,
    errorMessage: false
  }

  componentDidMount() {
    const { query } = this.props.match.params
    this.fetchAlbum(query)
  }

  componentWillReceiveProps(props) {
    const { query } = props.match.params
    this.fetchAlbum(query)
  }

  fetchAlbum = (query) => {
    this.setState({isLoading: true})
    const spotify = new SpotifyWrapper({
      token: TOKEN
    })
    spotify.search.albums(query)
    .then(data => {
      this.setState({isLoading: false})
      if (data.error) {
        this.setState({errorMessage: data.error.message})
      }else {
        this.setState({albums: data.albums.items})
      }
    })
  }

  render() {
    const { query } = this.props.match.params
    const { albums, isLoading, errorMessage } = this.state
    return (
      <Container>
        {isLoading && <Loader/> }
        <Search option={'album'} query={query} />
        {errorMessage && <Alert>{errorMessage}</Alert>}
        <List items={albums} />
      </Container>
    )
  }
}


export default withRouter(Album)
