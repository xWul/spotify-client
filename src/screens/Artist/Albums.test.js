import React from 'react'
import renderer from 'react-test-renderer'
import mockFetch from '../../api/mockFetch'
import Album from './Albums'

describe('Album Artist', function () {
  test('Artists Album renderer without crashing', () => {
    fetch = mockFetch({})
    const component = renderer.create(
      <Album />,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

})
