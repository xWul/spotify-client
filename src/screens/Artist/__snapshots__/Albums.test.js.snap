// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album Artist Artists Album renderer without crashing 1`] = `
<div
  className="sc-bdVaJa bxIkKu"
>
  <span
    className="sc-bwzfXH iccCXK"
    onClick={undefined}
  >
    Close
  </span>
  <ul
    className="sc-htpNat daVAfG"
  />
  <p
    className="sc-ifAKCX bnRFRw"
  >
    Loading...
  </p>
</div>
`;
